jax/_src/lax/lax.py

```diff
@@ -1116,17 +1116,18 @@ def iota(dtype: DType, size: int) -> Array:
   <https://www.tensorflow.org/xla/operation_semantics#iota>`_
   operator.
   """
-  dtype = dtypes.canonicalize_dtype(dtype)
-  size, = canonicalize_shape((size,))
-  return iota_p.bind(dtype=dtype, shape=(size,), dimension=0)
+  return broadcasted_iota(dtype, (size,), 0)
 
 def broadcasted_iota(dtype: DType, shape: Shape, dimension: int) -> Array:
   """Convenience wrapper around ``iota``."""
   dtype = dtypes.canonicalize_dtype(dtype)
   shape = canonicalize_shape(shape)
+  dynamic_shape = [d for d in shape if isinstance(d, core.Tracer)]
```

Comment:
This feels like it'll  be a pattern of sorts (decoupling a shape into static/dynamic portions and then stitching it back together in the impl).

It could be factored out into a pair of helper functions (one that splits up the shape and one that stitches it back together). Not sure if it's premature to do that though.

Reply:
It's a good idea! You're totally right. Actually @gnecula did that in #9426, so we should land that. I could copy them into this PR, but I figured we might land that one imminently.

Reply:
Okay then could you leave a TODO?

Reply:
Good idea, will do.